build_tools/cmake/rebuild.sh

```diff
@@ -35,11 +35,5 @@ else
   mkdir build
 fi
 cd build
-"$CMAKE_BIN" -G Ninja -DCMAKE_BUILD_TYPE=RelWithDebInfo \
-                      -DIREE_BUILD_COMPILER=ON \
-                      -DIREE_BUILD_TESTS=ON \
-                      -DIREE_BUILD_SAMPLES=OFF \
```

Comment:
Two tests are [failing on CMake](https://source.cloud.google.com/results/invocations/4d1b2699-561a-4a8b-a5d5-79a0dba1fd42/targets/iree%2Fgcp_ubuntu%2Fcmake%2Flinux%2Fx86-swiftshader%2Fpresubmit/log) but [passing on Bazel](https://source.cloud.google.com/results/invocations/c5e1e01e-291d-4923-a0a2-760482a42aa0/targets/iree%2Fgcp_ubuntu%2Fbazel%2Flinux%2Fx86-swiftshader%2Fcore%2Fpresubmit/log)?

```
RUNNING TEST: custom-opt -split-input-file /tmpfs/src/github/iree/iree/samples/custom_modules/dialect/test/custom_ops.mlir | custom-opt -split-input-file | IreeFileCheck /tmpfs/src/github/iree/iree/samples/custom_modules/dialect/test/custom_ops.mlir
----------------
/tmpfs/src/github/iree/iree/tools/run_lit.sh: line 115: IreeFileCheck: command not found
```

I'm trying to sort out why...

Reply:
Probably because Bazel is stricter about tracking deps and these ones aren't properly declared somehow. I think I've seen this as a flakey failure before.

Reply:
The direct deps are simple and appear correct: https://github.com/google/iree/blob/6ac002385f7cc6e4a3c997f84591e613c8238fc7/iree/samples/custom_modules/dialect/test/CMakeLists.txt#L13-L22

Seems like some part of https://github.com/google/iree/blob/main/build_tools/cmake/iree_lit_test.cmake isn't handling the current working directory in the same way as Bazel when running the tests (indirectly through several scripts). Other lit tests using tools like `iree-opt` have their tools at `iree/tools/`, while these have `custom-opt` at `iree/samples/custom_modules/dialect/`

Reply:
It looks like the `iree/tools` source directory is on the path:

> `EXPLICIT_PATH=/tmpfs/src/github/iree/iree/tools:/tmpfs/src/github/iree/build/iree/samples/custom_modules/dialect:`

but there wouldn't be an IreeFileCheck there, only IreeFileCheck.sh. I suspect the condition in https://github.com/google/iree/blob/6ac002385f7cc6e4a3c997f84591e613c8238fc7/build_tools/cmake/iree_lit_test.cmake#L66-L67 doesn't work. Everything else would get iree/tools on the path because of another dep

Reply:
I've hacked at this for a while and made little progress. I can tell that if I change `IreeFileCheck` to `IreeFileCheck.sh` in the test files, the tests get further.

Adding a `DEPENDS` on `FileCheck` here: https://github.com/google/iree/blob/0c2f9e8b847109f6ec0186a34fad3ab5c135ea48/iree/tools/CMakeLists.txt#L388-L395

or as a `DATA` dep here: https://github.com/google/iree/blob/0c2f9e8b847109f6ec0186a34fad3ab5c135ea48/iree/samples/custom_modules/dialect/test/CMakeLists.txt#L13-L22

then helps find `FileCheck` (used by `IreeFileCheck`)

I'm pretty confused tracing through the various layers of scripts, build system configuration, and platform specific details here. Seems like https://github.com/google/iree/issues/818 would help.

Reply:
You can't have comments between line continuations in bash :-P

Reply:
(the comment here was in reference to a separate issue on the same line. Actually doesn't make sense in this thread). The issue is as I described above. The source directory was getting added to the path instead of the binary directory, so just fixed that with [`fdbd553`](https://github.com/google/iree/pull/5197/commits/fdbd553500d5a897a51468334aaf2a83e96aecaf). I fiddled with trying to define a custom target for IreeFileCheck that would yield a `TARGET_FILE`, and got something that *works* though is pretty gross. And I'm not sure how it would actually work with dependency tracking

```cmake
  add_custom_command(
    OUTPUT IreeFileCheck
    COMMAND ${CMAKE_COMMAND} -E create_symlink ${CMAKE_CURRENT_SOURCE_DIR}/IreeFileCheck.sh IreeFileCheck
  )
  add_custom_target(_IreeFileCheck_Driver ALL
    DEPENDS IreeFileCheck
  )
  add_executable(iree_tools_IreeFileCheck IMPORTED GLOBAL)
  set_property(TARGET iree_tools_IreeFileCheck PROPERTY IMPORTED_LOCATION "${CMAKE_CURRENT_BINARY_DIR}/IreeFileCheck")
```

This allows removing the special casing in iree_lit_test.cmake

Reply:
That looks reasonable to me, actually. Removing the special case seems worth it.

Reply:
I think the issue is that the dependency information is lost, so I think depping on iree_tools_IreeFileCheck will no longer cause the custom command to run

Reply:
But maybe `add_depencencies` on the executable will work. I wasn't sure because it's imported, but let's see

Reply:
Ok I am once again deep in CMake confusion, but our current lit tests don't have targets you can build, so the whole dependency situation is tenuous anyway. I'll try out this fix

Reply:
Well adding an `add_dependencies(iree_tools_IreeFileCheck _IreeFileCheck_Driver)` causes a dependency cycle?!

```
CMake Error: The inter-target dependency graph contains the following strongly connected component (cycle):
  "_IreeFileCheck_Driver" of type UTILITY
    depends on "_IreeFileCheck_Driver" (strong)
The component contains at least one cycle consisting of strong dependencies (created by add_dependencies) that cannot be broken.
```

I think this has to do with

> Dependencies added to an imported target or an interface library are followed transitively in its place since the target itself does not build.

https://cmake.org/cmake/help/latest/command/add_dependencies.html

although I confess I don't actually understand what that sentence means...